### PR TITLE
Add Flashduty notification

### PR DESCRIPTION
**Pull request by guguji5**
Tick the checkbox if you understand [x]: 
- [x] I have read and understand the pull request rules.

# Description

Add notification provider of Flashduty (which is a Saas monitor noise reduction platform in China.)

## Type of change

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

### CODE REVIEW

File: server/notification-providers/flashduty.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+const { UP, DOWN, getMonitorRelativeURL } = require("../../src/util");
+const { setting } = require("../util-server");
+let successMessage = "Sent Successfully.";
+
+class FlashDuty extends NotificationProvider {
+    name = "FlashDuty";
+
+    /**
+     * @inheritdoc
+     */
```
Comment (CommanderStorm) [inline server/notification-providers/flashduty.js:12]:
```suggestion
```

File: server/notification-providers/flashduty.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        if (monitorInfo.type === "port") {
+            monitorUrl = monitorInfo.hostname;
+            if (monitorInfo.port) {
+                monitorUrl += ":" + monitorInfo.port;
+            }
+        } else if (monitorInfo.hostname != null) {
+            monitorUrl = monitorInfo.hostname;
+        } else {
+            monitorUrl = monitorInfo.url;
+        }
```
Comment (CommanderStorm) [inline server/notification-providers/flashduty.js:72]: Extracting this into another method might increase readability

**Review by CommanderStorm — CHANGES_REQUESTED** [review]

**Comment by CommanderStorm** [comment]
Could you also add a screenshot of the notification provider working? ^^
(This is standard procedure for regional notification providers, as testing them is sadly basically impossible due to region locks and language bariers)

File: src/components/notifications/FlashDuty.vue [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<template>
+    <div class="mb-3">
+        <label for="flashduty-integration-url" class="form-label">{{ $t("Integration URL") }}</label>
+        <input id="flashduty-integration-url" v-model="$parent.notification.flashdutyIntegrationUrl" type="text" class="form-control" autocomplete="false">
```
[the remark this answers is not in the document]
Comment (guguji5) [reply src/components/notifications/FlashDuty.vue]: Done as your suggestion

File: src/components/notifications/FlashDuty.vue [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    </div>
+    <div class="mb-3">
+        <label for="flashduty-priority" class="form-label">{{ $t("Priority") }}</label>
+        <select id="flashduty-priority" v-model="$parent.notification.flashdutyPriority" class="form-select">
```
[the remark this answers is not in the document]
Comment (guguji5) [reply src/components/notifications/FlashDuty.vue]: done

File: server/notification-providers/flashduty.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+class FlashDuty extends NotificationProvider {
+    name = "FlashDuty";
+
+    /**
+     * @inheritdoc
+     */
+    async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+        try {
+            if (heartbeatJSON == null) {
```
[the remark this answers is not in the document]
Comment (guguji5) [reply server/notification-providers/flashduty.js:12]: I send msg to replace the hardcode string.

File: server/notification-providers/flashduty.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            monitorUrl = monitorInfo.url;
+        }
+        const options = {
+            method: "POST",
+            url: notification.flashdutyIntegrationUrl,
+            headers: { "Content-Type": "application/json" },
+            data: {
+                description: `[${title}] [${monitorInfo.name}] ${body}`,
+                title,
+                event_status: notification.flashdutyPriority || "Warning",
```
[the remark this answers is not in the document]
Comment (guguji5) [reply server/notification-providers/flashduty.js]: I use Info as default value in frontend and server

File: src/components/NotificationDialog.vue [hunk 7]
```diff
@@ -157,6 +157,7 @@ export default {
                 "AliyunSMS": "AliyunSMS (阿里云短信服务)",
                 "DingDing": "DingDing (钉钉自定义机器人)",
                 "Feishu": "Feishu (飞书)",
+                "FlashDuty": "FlashDuty(快猫星云)",
```
Comment (chakflying) [inline src/components/NotificationDialog.vue:160]:
```suggestion
                "FlashDuty": "FlashDuty (快猫星云)",
```

**Review by chakflying — CHANGES_REQUESTED** [review]
Missing space.

**Comment by guguji5** [comment]
> Could you also add a screenshot of the notification provider working? ^^ (This is standard procedure for regional notification providers, as testing them is sadly basically impossible due to region locks and language bariers)

Because of the network, may I use the screenshot hosted by imgur?  (Currently, most of Flashduty users are in China, so the screenshot has some Chinese words.)

File: src/components/notifications/FlashDuty.vue [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+            <option value="Info" selected="selected">Info</option>
+            <option value="Warning">Warning</option>
+            <option value="Critical">Critical</option>
```
[the remark this answers is not in the document]
Comment (guguji5) [reply src/components/notifications/FlashDuty.vue]: Thanks for your suggestion. But the 'Info Warning Critical Ok' are key word in Alert Status of Flashduty (Even display as that in Chinese language system),So I think it doesn't need translation.

File: src/components/NotificationDialog.vue (hunk 7 above)
Comment (guguji5) [reply src/components/NotificationDialog.vue]: done

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
> Because of the network, may I use the screenshot hosted by imgur?

If uploading to github is not an option, that is also fine. 

> Currently, most of Flashduty users are in China, so the screenshot has some Chinese words.

That is fine, both Nelson and Louis can read Chinese ^^

File: server/notification-providers/flashduty.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+     * Send the message
+     * @param {BeanModel} notification Message title
+     * @param {string} title Message
+     * @param {string} body Message
+     * @param {Object} monitorInfo Monitor details
+     * @param {string} eventStatus Monitor status (Info, Warning, Critical, Ok)
+     * @returns {string}
+     */
+    async postNotification(notification, title, body, monitorInfo, eventStatus) {
+        const genMonitorUrl = (monitorInfo) => {
```
[the remark this answers is not in the document]
Comment (guguji5) [reply server/notification-providers/flashduty.js:60]: of course, I will make a refactor here later.

File: server/notification-providers/flashduty.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            if (heartbeatJSON.status === DOWN) {
+                const title = "Uptime Kuma Monitor 🔴 Down";
+                return this.postNotification(notification, title, heartbeatJSON.msg, monitorJSON, notification.flashdutySeverity);
+            }
+        } catch (error) {
+            this.throwGeneralAxiosError(error);
+        }
+    }
+
+    genMonitorUrl(monitorInfo) {
```
Comment (CommanderStorm) [inline server/notification-providers/flashduty.js:36]: I am unsure if the undefined case works correctly with the notification provider
See https://github.com/louislam/uptime-kuma/issues/3274 for an instance of what is my concern ^^
(also unsure if url were null or undefined in this case…)

```suggestion
    /**
     * generate a monitor url from the monitors infomation
     * @param {Object} monitorInfo Monitor details
     * @returns {string|undefined} url linking to the monitor
     */
    genMonitorUrl(monitorInfo) {
```

**Comment by guguji5** [comment]
> about the screenshots: Could you attach a screenshot showing how the `Test` button looks when pressed? (I am a bit concearned that specifing it as a `url` => using it as a `resource` in `postNotification` might not work)

Yep, you are right. there is a bug when fire a 'Test' request. I fix it by adding '?' before the `monitorInfo` (which means when the  monitorInfo is undefined,  abandon the logic behind and return undefined.)

File: server/notification-providers/flashduty.js (hunk 10 above)
Comment (guguji5) [reply server/notification-providers/flashduty.js:42]: When Testing, the url is undefined, but when the monitor up or down ,the url will be got successfully. as the below screenshot. the param description has been appended.

**Review by CommanderStorm — APPROVED** [review]

**Review by chakflying — APPROVED** [review]
I think including `zh-HK` and `zh-TW` translation is unnecessary. Otherwise looks okay.

**Comment by guguji5** [comment]
@CommanderStorm is it ready to merge this PR ? or anything to improve ?

**Comment by CommanderStorm** [comment]
As indicated by my aproval-review:
from my point, no, but only louis vote counts. See our contribution guide: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md
The point of reviews from me and Nelson is to reduce unnessesary go-arounds from Louis, not to give any indicaor of timeline/mergability/...

**Comment by guguji5** [comment]
> As indicated by my aproval-review: from my point, no, but only louis vote counts. See our contribution guide: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md The point of reviews from me and Nelson is to reduce unnessesary go-arounds from Louis, not to give any indicaor of timeline/mergability/...

thank you for your reply. I will wait louis's feedback.

**Comment by louislam** [comment]
> I think including `zh-HK` and `zh-TW` translation is unnecessary. Otherwise looks okay.

Same thought, just need to change zh-CN.json, as the others are using Traditional Chinese.

Other code changes look good to me.

**Comment by guguji5** [comment]
> > I think including `zh-HK` and `zh-TW` translation is unnecessary. Otherwise looks okay.
> 
> Same thought, just need to change zh-CN.json, as the others are using Traditional Chinese.
> 
> Other code changes look good to me.

@chakflying @louislam ok, I revert the changes in zh-HK and zh-TW and fix the conflicts in en.json

**Pull request merged**